### DISCUSSION TITLE
I've made a trade plugin 'SpaceCrate' - Merchants, check it out and give me your thoughts.

### DISCUSSION THREAD

**Discussion by adegans** [endless-sky/endless-sky#7824] [Show and tell]
Hi guys,

I'm a fairly new player and spent the last few days scripting a trade oriented plugin called SpaceCrate for Endless Sky 0.9.16.1.
I would like your thoughts on it in its current form. Perhaps ideas for more stuff to add in future releases.
If you're into trade you might find this plugin fun for extra content and missions.

Also I'm looking to find out how to get it listed on the Plugins page. If anyone knows who to talk to to get that done that would be great.

**Download:**
https://github.com/adegans/SpaceCrate

**From the readme:**
SpaceCrate is a trade mod focused on lucrative deals from a trade hub perspective.
I've added a rich planet called X-283N1 whose people are obsessed with trade and exports.
An implied silly 'narrative' exists appointing Arnan (me) as its CEO running the place. 
The people on X-283N1 love trade and exports so much that they have no time for petty 
politics. All they want is to ship stuff to the galaxy. And they're willing to pay.

The riches this implies of-course attracts workers from all over the galaxy (passenger jobs 
in most human systems) this in turn results in shipments of all sizes being produced ready 
for export (cargo jobs). Higher paying and more challenging jobs (bulk shipments/rush jobs) 
are unlocked as your combat rating goes up. With all jobs accessible once you reach level 10.

You will not find fighting jobs or other combat related activities in this plugin as that 
is not what this mod is for. Instead the mod relies on already available mechanics to create
a (pirate) threat for your growing merchant fleet. The planet does have a sizable defense 
force that will attack you like a bunch of mad pirates should you contest the planet.

You'll see CEO Arnan fly in and around Alpha Centauri in his ship 'StarGoose' encouraging 
you to visit his planet and make some money if you hail him.

SpaceCrate is a bit of an experiment with how the plugin system works.

**Comment by Zitchas** [comment]
Neat! We can always use more plugins.

To get it listed on the plug-in page you will need to make a pull request to the manifest directory that adds a manifest file named after your plugin.

Here's my plugin manifest, as a an example of what it looks like:
https://github.com/endless-sky/endless-sky-plugins/blob/master/manifests/Z%20Lampyrid%20Start.yaml

Here's the repository link:
https://github.com/endless-sky/endless-sky-plugins

Here's the description of the manifest contents:
https://github.com/endless-sky/rfcs/blob/main/rfcs/0001-plugin-index.md

**Reply by adegans** [reply]
I can create the YAML. But as I mentioned I don't use github - Assuming the pull request stuff applies to that...
I just have an account so I can post on forums 😉

**Reply by Zitchas** [reply]
Yes, you'd need to use a little github to do that; although one of our experts might be able to provide a walkthrough. I doubt you are the only person who doesn't know how, so having a walkthrough would be a useful thing.

**Reply by adegans** [reply]
Well I put the YAML in the repo. Can't figure out pull requests... Nothing seems to work or be clickable to lets me share or 'send' a file to the plugin repo. So I give up... If you want it - grab it here https://github.com/adegans/SpaceCrate/blob/main/spacecrate.yaml

**Reply by quyykk** [reply]
I believe the autoupdate URL needs to include `$version` instead of the actual version number but other than that LGTM. @MCOfficer

**Reply by adegans** [reply]
Hmm - I've updated it, should be good now.

**Reply by quyykk** [reply]
If you want I can create the PR for you. Alternatively, here's an image on how to do it yourself:

![image](https://user-images.githubusercontent.com/85879619/206877970-693af71d-4608-4215-b33f-f74146b0b34e.png)

You'll then see your file, and at the bottom you have the option to select "create new PR".

**Reply by adegans** [reply]
Because why would we use the Pull Requests menu to do  a pull request ;)  I'll give it a go. Thanks!!

**Reply by adegans** [reply]
No Joy:
<img width="896" alt="Screenshot 2022-12-10 at 15 43 52" src="https://user-images.githubusercontent.com/9438487/206878103-50a40ad9-8e70-4828-a6ec-5f53f835c43a.png">

<img width="619" alt="Screenshot 2022-12-10 at 15 43 56" src="https://user-images.githubusercontent.com/9438487/206878105-5ce5911b-3524-4a61-9980-6d71d6271c31.png">

**Reply by quyykk** [reply]
Maybe "create new file" then?

**Reply by adegans** [reply]
That works, but then my 'own' file is redundant? The one in my repo.

**Reply by quyykk** [reply]
Yeah, you only need the manifest in the general plugins repo.

**Reply by adegans** [reply]
Right, I've done the thing - I think. Should show up now.
I'll remove my own copy. Cheers.

**Reply by adegans** [reply]
Do I need to update the yaml in the plugins repo or something when I release a new version?
I'll make a release in the SpaceCrate repo, does the game autoupdate plugins? Or how does that work?

**Reply by quyykk** [reply]
> Do I need to update the yaml in the plugins repo or something when I release a new version?

No, the auto updater will take care of that for you.

> ll make a release in the SpaceCrate repo, does the game autoupdate plugins?

Yes, just make sure the new version is actually higher than the previous ones.

**Reply by quyykk** [reply]
Oh wait I misread, the game itself doesn't update anything.

**Reply by adegans** [reply]
Then how do casual players update plugins or even know about updates?

**Reply by quyykk** [reply]
There's [ESL2](https://github.com/EndlessSkyCommunity/ESLauncher2) that people use. In the future (hopefully next year :crossed_fingers:), the game will be able to download and update plugins.

**Reply by adegans** [reply]
Hmm, hopefully at least a notification thing gets added soon.

**Comment by adegans** [comment]


**Reply by Zitchas** [reply]
It takes a bit of work, yeah. Took me many months when I got involved with Endless Sky to get comfortable with GitHub.

edit: And I'm still not completely comfortable with it, especially the advanced stuff. But enough to be productive!

**Reply by adegans** [reply]
Did you try/test SpaceCrate yet? 🤣

**Reply by Zitchas** [reply]
Sorry, no. Working on some code to fix a bug; and got some RL stuff to take care of.

**Comment by adegans** [comment]
Anyone tried the plugin yet?
What do you think?
Any issues with it?

**Reply by quyykk** [reply]
I've posted this discussion to the Discord, so you'll get a bit more exposure :smile:

**Comment by adegans** [comment]
Version 0.3 is pretty much done.

**Comment by adegans** [comment]
Version 0.3 is out - Grab it here: https://github.com/adegans/SpaceCrate/releases/tag/v0.3
Check readme to see all changes, or play and find out as you do missions.

**Comment by adegans** [comment]
SpaceCrate 0.4 is available now https://github.com/adegans/SpaceCrate/releases/tag/v0.4
Primarily fixed many spelling errors and added a side story, story line.
Check the readme for a changelog.

**Reply by 1010todd** [reply]
Finally got a chance to play around with it. First thought is their job board is very empty (brand new pilot), I think they need some generic jobs going between them and other planets like normal government do that don't have combat rating requirement, this might also help some player who go in blind find the planet. Their government color is also very bright, for color in ES data we use range between 0.-1. not 0-255, anything over 1 will make the color brighter (which could also be intentional to make them stand out more).

Also, I happen to have a plugin that add several human freighters (Mega Freight) and maybe we could do a little cross-plugin compatibility by having Spacecrate merchant fleet also use these freighters as well if both plugin is installed if you're fine with that. (You don't have to do anything as I can check and add the ships to the fleet from my side.) Just thought it'd be interesting to see more ES plugins have cross-compatibility, even just a little.

**Reply by adegans** [reply]
Sure, we can do some collab on that. I considered adding SpaceCrate specific ships, but didn't feel like designing sprites for it for now because that, for me, would take quite a lot of time and effort compared to how many people actually play the game.
And since I'm not adding an actual race, just a bunch of trade hungry humans it seemed a bit much to create a whole line of ships for it. Plus I'm not a graphics guy.
But if you have some cool ships, sure, go for it 👍 Got any pictures? Or what plugin do I install to see them?

If you want to discuss this outside of Github hit me up through one of these  - https://www.arnan.me/links.html

I'll look at the color. I never was overly excited about the yellow and only kept it around because no other faction had a color like it.
Suggestions welcome.

And indeed, if you're not working for SpaceCrate, missions are a bit harder to find and are just like other missions. This is by design. Once you're "in" they'll be plenty and pay a lot better. But they also get a bit more challenging or interesting as they require bigger ships/fleets to complete. You think there are too few? Suggestions?

On another note - Since you seem knowledgable about the game. I'm having some trouble with the trigger for the employment storyline. As-in it doesn't reliably get triggered. Any thoughts on that? It should when landing on the SpaceCrate planet but I can't quite get it to work - Not sure why.

I could just have it show up on the job board, but that's not consistent with how this kind of mission is triggered in the game I think.

Here is the mission spec for it:
```
mission "spacecrate_story_signup_1"
	name "SpaceCrate - Employment opportunity"
	description "Travel to <destination> where you'll pick up raw materials and join a merchant fleet."
	landing
	autosave
	minor
	source "X-283N1"
	destination "Antipode"
	to offer
		"spacecrate_employee" == 0
		"spacecrate_career_reset" == 0
		"reputation: SpaceCrate" > 5000
		"cargo space" >= 450
	on offer
[... the rest...]
```

Perhaps you can spot the error...

**Reply by 1010todd** [reply]
I don't have gallery of it for browsing but the plugin is here: https://github.com/1010todd/Mega-Freight . Note it's not exactly balanced against vanilla (or at all in some area).

It's not just few mission, there is nothing in the job board if you don't have little to no combat rep (what I would've expect from primarily trading playthrough). It's just a little odd to see a completely empty job board so I think there should still be a few generic one just like what you can get elsewhere that doesn't pay extra or have any requirement beside the space to complete it.

As for the mission, if you mean it doesn't pop up everytime you expect it to be then maybe it's the `minor` tag, it means that other missions offering at the time will block the mission with `minor`, that also includes invisible missions so it could be some invisible mission blocking it.

**Reply by adegans** [reply]
Those are some nice ships you got there.

No jobs is not right, I’ll look into that.

**Comment by adegans** [comment]
Finally fixed the storyline trigger (I think). And tweaked a few other things in version 0.4.1.
https://github.com/adegans/SpaceCrate/releases/tag/v0.4.1
